src/renderers/shared/reconciler/ReactDataTracker.js

```diff
@@ -0,0 +1,144 @@
+        var oldListeners = ReactDataTracker.listeners.get(oldDataSources[index]);
+        oldListeners.splice(oldListeners.indexOf(component), 1);
+        oldDataSources.splice(index, 1);
+        index--;
+      }
+    }
+    for (index = 0; index < newDataSources.length; index++) {
+      if (oldDataSources.indexOf(newDataSources[index]) === -1) {
+        if (!ReactDataTracker.listeners.has(newDataSources[index])) {
+          ReactDataTracker.listeners.set(newDataSources[index], []);
+        }
+        ReactDataTracker.listeners.get(newDataSources[index]).push(component);
+        ReactDataTracker.dataSources.get(component).push(newDataSources[index]);
+      }
+    }
+  },
+
+  startRead: function(entity) {
+    if (ReactDataTracker.activeReaders === undefined) {
+      ReactDataTracker.activeReaders = 0;
+    }
+    ReactDataTracker.activeReaders++;
+  },
+
+  endRead: function(entity) {
+    if (ReactDataTracker.currentContext !== undefined && ReactDataTracker.currentContext.indexOf(entity) === -1) {
+      ReactDataTracker.currentContext.push(entity);
+    }
+    ReactDataTracker.activeReaders--;
+    if (ReactDataTracker.activeReaders < 0) {
+      throw new Error('Number of active readers dropped below zero');
+    }
+  },
+
+  startWrite: function(entity) {
+    if (ReactDataTracker.writers === undefined) {
+      ReactDataTracker.writers = [];
+    }
+    if (ReactDataTracker.writers.indexOf(entity) === -1) {
+      ReactDataTracker.writers.push(entity);
+    }
+    if (ReactDataTracker.activeWriters === undefined) {
+      ReactDataTracker.activeWriters = 0;
+    }
+    ReactDataTracker.activeWriters++;
+  },
+
+  endWrite: function(entity) {
+    if (ReactDataTracker.activeWriters === undefined) {
+      throw new Error('Can not end write without starting write');
+    }
+    if (ReactDataTracker.writers.indexOf(entity) === -1) {
+      throw new Error('Can not end write without starting write');
+    }
+    ReactDataTracker.activeWriters--;
+
+    if (ReactDataTracker.activeWriters === 0) {
+      // for each writer that wrote during this batch
+      var componentsToNotify = [];
+      for (var writerIndex = 0; writerIndex < ReactDataTracker.writers.length; writerIndex++) {
+        var writer = ReactDataTracker.writers[writerIndex];
+        if (ReactDataTracker.listeners === undefined) {
+          continue;
+        }
+        if (!ReactDataTracker.listeners.has(writer)) {
+          continue;
+        }
+        var listenersList = ReactDataTracker.listeners.get(writer);
+        for (var index = 0; index < listenersList.length; index++) {
+          if (componentsToNotify.indexOf(listenersList[index]) === -1) {
+            componentsToNotify.push(listenersList[index]);
+          }
+        }
+      }
+
+      for (var componentIndex = 0; componentIndex < componentsToNotify.length; componentIndex++) {
+        componentsToNotify[componentIndex].setState({});
```

Comment:
Maybe use `ReactUpdateQueue.enqueueForceUpdate(this)`? This seems like it wouldn't work with components with strict `shouldComponentUpdate`, and will throw on non-Component-descending components.